pkg/storage/tree/tree.go

```diff
@@ -109,17 +108,7 @@ func prependInt(s []int, x int) []int {
 }
 
 func (t *Tree) String() string {
-	t.RLock()
-	defer t.RUnlock()
-
-	res := ""
-	t.Iterate(func(k []byte, v uint64) {
-		if v > 0 {
-			res += fmt.Sprintf("%q %d\n", k[2:], v)
-		}
-	})
-
-	return res
+	return t.Collapsed()
```

Comment:
I wonder why `String()` is built on top of `Collapsed()`... it may be simpler to just have a single implementation under `String` and just call `String()` from the controller, right? No need to have 2 different methods for the same thing.

I'm also curious about why the original implementation has been discarded and a new iteration function has been created. Is there some problem with the original code and, if there is, could the old code be replaced by the new one instead of having two similar methods?

Reply:
Yes I thought this might raise some eyebrows. The way I thought about this is , collapsed is a standardized format. String incidentally uses something very close to this format, but it could very well be something else. This makes that independence explicit.
Why not use the original Iterate - It returns a semicolon delimited list of frames , which is not very flexible. So you end up having to join then the output split to manipulates the frames, and it has other idiosyncrasies like needing to check for len(name) > 2 for empty stacks etc. It was probably written to Stringify the tree and hasn't grown out of it. I don't want to change it too much without having a firm grasp on how it's used in the ingestion code.